examples/research_projects/jax-projects/model parallel/partitions.py

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2021 The Google Research Authors and The HuggingFace Team All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for constructing PyTrees of PartitionSpecs."""
+
+# utils adapted from https://github.com/google-research/google-research/blob/master/flax_models/t5x/partitions.py
+
+import re
+
+from flax.core.frozen_dict import freeze
+from flax.traverse_util import flatten_dict, unflatten_dict
+from jax.experimental import PartitionSpec as P
+
+
+# Sentinels
+_unmatched = object()
+
+# For specifying empty leaf dict `{}`
+empty_dict = object()
+
+
+def _match(qs, ks):
+    """Return True if regexes in qs match any window of strings in tuple ks."""
+    # compile regexes and force complete match
+    qts = tuple(map(lambda x: re.compile(x + "$"), qs))
+    for i in range(len(ks) - len(qs) + 1):
+        matches = [x.match(y) for x, y in zip(qts, ks[i:])]
+        if matches and all(matches):
+            return True
+    return False
+
+
+def _replacement_rules(rules):
+    def replace(key, val):
+        for rule, replacement in rules:
+            if _match(rule, key):
+                return replacement
+        return val
+
+    return replace
+
+
+# PartitionSpec for GPTNeo
+# replicate the hidden dim and shard feed-forward and head dim
+def _get_partition_rules():
```

Comment:
For the future:

Do you think we'll always need model-specific partitions? This might make it a bit difficult to implement for all models, no? I guess we can't just assign the correct `P("mp", None)` depending on the last word "kernel" or "embedding"  since they switch...

Maybe it's somehow possible to trace a model (similar to how init does in to give us the params) to automatically build this map?

Reply:
That's the idea. With lazy initialization, we get the `PyTree` of params which describes all param names and shapes. 
With that, it should be possible to build a default spec for every model. 

We can already do this with `jax.eval_shape`. cf
https://github.com/huggingface/transformers/blob/399258357bad74186f248a373af8012f2b0203b7/examples/research_projects/jax-projects/model%20parallel/run_clm_mp.py#L471

Reply:
This is typically done using using various load balancing methods, e.g. deepspeed pipe has:
https://www.deepspeed.ai/tutorials/pipeline/#load-balancing-pipeline-modules
pytorch has these too but I can't find any mentions of these in their docs.

Is that what you're referring to, @patrickvonplaten